src/Microsoft.Azure.WebJobs.Logging/Internal/LogReader.cs

```diff
@@ -155,20 +184,186 @@ public LogReader(CloudTable table)
+                }
+                return d;
+            }
+
+            // Given an epoch, find the next one in physicalEpochs that it would fall into, using reverse-chronological order.              
+            // physicalEpochs - epochs for existing tables, sorted in ascending order.  This is generally short (expected under 50 elements)
+            // Returns a physical epoch (from the list passed in) that the requested epoch falls into 
+            // or -1 if there's no physical epoch below this.
+            private static long FindNext(long epoch, long[] physicalEpochs)
+            {
+                // start at most recent epoch and work backwards 
+                for (var i = physicalEpochs.Length - 1; i >= 0; i--)
+                {
+
+                    if (epoch > physicalEpochs[i])
+                    {
+                        epoch = physicalEpochs[i];
+                        return epoch;
+                    }
+
+                    if (epoch == physicalEpochs[i])
+                    {
+                        return epoch;
+                    }
+                }
+
+                return -1; // done 
+            }
+
+            // Helper workaround for this bug: 
+            // https://github.com/projectkudu/AzureFunctionsPortal/issues/634 
+            // Portal needs to be updated to use continuation tokens. Once it is, we can remove this shim
+            // and let the portal drive the enumeration directly. 
+            public async Task<Segment<TElement>> SafeExecuteQuerySegmentedAsync<TElement>(
+                TableQuery<TElement> rangeQuery,
+                DateTime startTime, // Range for query 
+                DateTime endTime
+            ) where TElement : ITableEntity, new()
+            {
+                List<TElement> list = new List<TElement>();
+
+                string continuationToken = null;
+                while (true)
+                {
+                    var segment = await SafeExecuteQuerySegmentedAsync(rangeQuery, startTime, endTime, continuationToken);
+
+                    list.AddRange(segment.Results);
+                    if (segment.ContinuationToken == null)
+                    {
+                        // Done!
+                        return new Segment<TElement>(list.ToArray(), null);
+                    }
+                    continuationToken = segment.ContinuationToken;
+                }
+            }
+
+            // Date range queries start with most recent (endTime) and then return entities in descending chronological order. 
+            public async Task<Segment<TElement>> SafeExecuteQuerySegmentedAsync<TElement>(
+                TableQuery<TElement> rangeQuery,
+                DateTime startTime, // Range for query 
+                DateTime endTime,
+                string continuationToken
+                ) where TElement : ITableEntity, new()
+            {
+                _tables = await InitAsync(_tableLookup);
```

Comment:
we do anyways; but you're right - this could be cleaner.  see next iteration

---
In reply to: [82880453](https://github.com/Azure/azure-webjobs-sdk/pull/862#discussion_r82880453) [](ancestors = 82880453)